google-cloud-bigquery/src/test/java/com/google/cloud/bigquery/BigQueryRetryHelperTest.java

```diff
@@ -0,0 +1,170 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.bigquery;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import com.google.api.client.googleapis.json.GoogleJsonError;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.client.http.HttpHeaders;
+import com.google.api.client.http.HttpResponseException;
+import com.google.api.core.NanoClock;
+import com.google.api.gax.retrying.BasicResultRetryAlgorithm;
+import com.google.api.gax.retrying.RetrySettings;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.io.IOException;
+import java.net.SocketException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.junit.Test;
+import org.threeten.bp.Duration;
+
+public class BigQueryRetryHelperTest {
+
+  @Test
+  public void testRetriesMultithread() throws ExecutionException, InterruptedException {
+    BigQueryRetryConfig config =
+        BigQueryRetryConfig.newBuilder().retryOnMessage("asdf").retryOnRegEx(".*dfdf.*").build();
+
+    ThreadFactory threadFactory =
+        new ThreadFactoryBuilder().setDaemon(true).setNameFormat("testing-%02d").build();
+
+    ListeningExecutorService exec =
+        MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(4, threadFactory));
+
+    int endInclusive = 16;
+    RetrySettings retrySettings =
+        RetrySettings.newBuilder()
+            .setMaxAttempts(6)
+            .setInitialRetryDelay(Duration.ofMillis(20))
+            .setRetryDelayMultiplier(1.25)
+            .setMaxRetryDelay(Duration.ofMinutes(1))
+            .build();
+    ListenableFuture<List<Result>> f =
+        Futures.allAsList(
+            IntStream.rangeClosed(1, endInclusive)
+                .mapToObj(i -> exec.submit(new ResultCallable(i, config, retrySettings)))
+                .collect(Collectors.toList()));
+
+    List<Result> results = f.get();
+
+    List<String> values = results.stream().map(r -> r.val).collect(Collectors.toList());
+    Map<String, List<String>> grouped =
+        values.stream().collect(Collectors.groupingBy(Function.identity()));
+    int sum = results.stream().mapToInt(r -> r.count).sum();
+
+    assertThat(values).hasSize(endInclusive);
```

Comment:
Can we simplify this test case to just show we are retrying with the correct maxAttempt and let's make maxAttempt a variable which is being asserted on. 

The idea here is to verify that we are not doing infinite retries and we should show that as clearly as possible.

Reply:
I've simplified this a bit and added some comments to explain what is being asserted.

These assertions are trying to ensure all of the retries are happening across all of the jobs and that all jobs ultimately succeed so I think we need to keep at least the two asserts I've left.